Changes in module 'nf-core/bakta/baktadbdownload'
--- modules/nf-core/bakta/baktadbdownload/main.nf
+++ modules/nf-core/bakta/baktadbdownload/main.nf
@@ -1,10 +1,10 @@
 process BAKTA_BAKTADBDOWNLOAD {
     label 'process_single'
 
-    conda "bioconda::bakta=1.8.2"
+    conda "bioconda::bakta=1.8.1"
     container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/bakta:1.8.2--pyhdfd78af_0' :
-        'biocontainers/bakta:1.8.2--pyhdfd78af_0' }"
+        'https://depot.galaxyproject.org/singularity/bakta:1.8.1--pyhdfd78af_0' :
+        'biocontainers/bakta:1.8.1--pyhdfd78af_0' }"
 
     output:
     path "db*"              , emit: db

************************************************************
