Changes in module 'nf-core/nanoplot'
--- modules/nf-core/nanoplot/main.nf
+++ modules/nf-core/nanoplot/main.nf
@@ -22,13 +22,15 @@
 
     script:
     def args = task.ext.args ?: ''
-    def input_file = ("$ontfile".endsWith(".fastq.gz")) ? "--fastq ${ontfile}" :
-        ("$ontfile".endsWith(".txt")) ? "--summary ${ontfile}" : ''
+    def prefix = task.ext.prefix ?: "${meta.id}"
+    def input_file = ("$ontfile".endsWith(".fastq.gz") || "$ontfile".endsWith(".fq.gz")) ? "--fastq ${ontfile}" :  ("$ontfile".endsWith(".txt")) ? "--summary ${ontfile}" : ''
     """
     NanoPlot \\
         $args \\
         -t $task.cpus \\
         $input_file
+
+    mv NanoStats.txt ${prefix}.txt
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
