Changes in module 'nf-core/racon'
--- modules/nf-core/racon/main.nf
+++ modules/nf-core/racon/main.nf
@@ -11,7 +11,7 @@
     tuple val(meta), path(reads), path(assembly), path(paf)
 
     output:
-    tuple val(meta), path('*_assembly_consensus.fasta.gz') , emit: improved_assembly
+    tuple val(meta), path('*.consensus.fasta.gz') , emit: improved_assembly
     path "versions.yml"          , emit: versions
 
     when:
@@ -26,9 +26,9 @@
         "${paf}" \\
         $args \\
         "${assembly}" > \\
-        ${prefix}_assembly_consensus.fasta
+        ${prefix}.consensus.fasta
 
-    gzip -n ${prefix}_assembly_consensus.fasta
+    gzip -n ${prefix}.consensus.fasta
 
     cat <<-END_VERSIONS > versions.yml
     "${task.process}":

************************************************************
